### PR TITLE
Reinstate local web-tree-sitter build

### PR DESCRIPTION
**Pull request by pokey**
Well that was short-lived 😅. I ran into a web-tree-sitter bug that is fixed on master but unreleased, so I reintroduced the local build machinery to allow us to use unreleased web-tree-sitter versions

It's worth looking at a [diff against the previous build machinery](https://github.com/cursorless-dev/vscode-parse-tree/compare/1a0ce328cc4d13161294c716fe1a900affe4996a...pokey/reinstate-local-web-tree-sitter-build) to see what's changed

- Partially reverts #68

### CODE REVIEW

**Comment by pokey** [comment]
@phillco it's worth giving this one a whirl to see if this works for you

**Comment by phillco** [comment]
This build fails for me, but I don't know if it's just an issue with my dependency setup. https://gist.github.com/phillco/30557f14452ce3b9649702c6b1aac396

FWIW, I get a different error on main:

```
~/s/c/vscode-parse-tree (0 ms)  (main|✔) > yarn
yarn install v1.22.19
warning ../../package.json: No license field
$ make web-tree-sitter
emcc: error: undefined exported symbol: "__ZNKSt3__220__vector_base_commonILb1EE20__throw_length_errorEv" [-Wundefined] [-Werror]
make: *** [vendor/web-tree-sitter/0.20.4/README.md] Error 1
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by wenkokke** [comment]
If you're already using pnpm, I'd suggest using their patchedDependencies feature:

https://pnpm.io/package_json

**Comment by pokey** [comment]
@phillco that issue you're seeing on this branch is due to Python version. Try building this branch with Python 3.10. There's some node gyp issue with Python 3.11

Not sure what's up with the error on main

**Comment by phillco** [comment]
Hm, still fails:

```
...
gyp info find Python using Python version 3.10.5 found at "/Users/phillco/.pyenv/versions/3.10.5/bin/python3"
gyp ERR! UNCAUGHT EXCEPTION
gyp ERR! stack TypeError: Cannot assign to read only property 'cflags' of object '#<Object>'
gyp ERR! stack     at createConfigFile (/Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/lib/configure.js:117:21)
gyp ERR! stack     at /Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/lib/configure.js:84:9
gyp ERR! stack     at FSReqCallback.oncomplete (node:fs:189:23)
gyp ERR! System Darwin 21.6.0
gyp ERR! command "/opt/homebrew/Cellar/node/20.8.0/bin/node" "/Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native/binding-v6.0.1-n-api-darwin-arm64/lzma_native.node" "--module_name=lzma_native" "--module_path=/Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native/binding-v6.0.1-n-api-darwin-arm64" "--napi_version=9" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v115"
gyp ERR! cwd /Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native
gyp ERR! node -v v20.8.0
gyp ERR! node-gyp -v v7.1.2
gyp ERR! Node-gyp failed to build your package.
```

File: Makefile [hunk 1]
```diff
@@ -36,3 +36,31 @@ parsers/tree-sitter-c-sharp.wasm: node_modules/tree-sitter-c-sharp/package.json
...
+$(addprefix $(WEB_TREE_SITTER_DIR)/,$(WEB_TREE_SITTER_FILES)): tree-sitter-version
+	@rm -rf $(MAKE_CACHE_DIR)/tree-sitter
+	@TREE_SITTER_VERSION=$(shell cat tree-sitter-version) ;\
+	git clone                                       \
+		-c advice.detachedHead=false --quiet           \
+		--depth=1 \
+		https://github.com/tree-sitter/tree-sitter.git \
+		$(MAKE_CACHE_DIR)/tree-sitter ;\
+	cd $(MAKE_CACHE_DIR)/tree-sitter && \
+	git fetch origin $$TREE_SITTER_VERSION && \
```
Comment (auscompgeek) [inline Makefile:60]: Do we need a `git clone`? Can we fetch a tarball archive instead?

**Comment by pokey** [comment]
> Hm, still fails:
> 
> ```
> ...
> gyp info find Python using Python version 3.10.5 found at "/Users/phillco/.pyenv/versions/3.10.5/bin/python3"
> gyp ERR! UNCAUGHT EXCEPTION
> gyp ERR! stack TypeError: Cannot assign to read only property 'cflags' of object '#<Object>'
> gyp ERR! stack     at createConfigFile (/Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/lib/configure.js:117:21)
> gyp ERR! stack     at /Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/lib/configure.js:84:9
> gyp ERR! stack     at FSReqCallback.oncomplete (node:fs:189:23)
> gyp ERR! System Darwin 21.6.0
> gyp ERR! command "/opt/homebrew/Cellar/node/20.8.0/bin/node" "/Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native/binding-v6.0.1-n-api-darwin-arm64/lzma_native.node" "--module_name=lzma_native" "--module_path=/Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native/binding-v6.0.1-n-api-darwin-arm64" "--napi_version=9" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v115"
> gyp ERR! cwd /Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native
> gyp ERR! node -v v20.8.0
> gyp ERR! node-gyp -v v7.1.2
> gyp ERR! Node-gyp failed to build your package.
> ```

looks like you're on a fairly old node-gyp version; try upgrading and see if that helps (`npm update -g node-gyp`)

**Comment by phillco** [comment]
```
~/s/c/vscode-parse-tree (14 ms)  (pokey/reinstate-local-web-tree-sitter-build|…) > npm update -g node-gyp

changed 1 package in 2s

155 packages are looking for funding
  run `npm fund` for details
```

```
gyp info using node-gyp@7.1.2
gyp info using node@20.8.0 | darwin | arm64
gyp info find Python using Python version 3.10.13 found at "/Users/phillco/.pyenv/versions/3.10.13/bin/python3"
gyp ERR! UNCAUGHT EXCEPTION
gyp ERR! stack TypeError: Cannot assign to read only property 'cflags' of object '#<Object>'
gyp ERR! stack     at createConfigFile (/Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/lib/configure.js:117:21)
gyp ERR! stack     at /Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/lib/configure.js:84:9
gyp ERR! stack     at FSReqCallback.oncomplete (node:fs:189:23)
gyp ERR! System Darwin 21.6.0
gyp ERR! command "/opt/homebrew/Cellar/node/20.8.0/bin/node" "/Users/phillco/src/c/vscode-parse-tree/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native/binding-v6.0.1-n-api-darwin-arm64/lzma_native.node" "--module_name=lzma_native" "--module_path=/Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native/binding-v6.0.1-n-api-darwin-arm64" "--napi_version=9" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v115"
gyp ERR! cwd /Users/phillco/src/c/vscode-parse-tree/node_modules/lzma-native
gyp ERR! node -v v20.8.0
gyp ERR! node-gyp -v v7.1.2
gyp ERR! Node-gyp failed to build your package.
gyp ERR! Try to update npm and/or node-gyp and if it does not help file an issue with the package author.
```

Is it using the right node-gyp?

```
~/s/c/vscode-parse-tree (120 ms)  (pokey/reinstate-local-web-tree-sitter-build|…) > node-gyp

  Usage: node-gyp <command> [options]

  where <command> is one of:
    - build - Invokes `make` and builds the module
    - clean - Removes any generated build files and the "out" dir
    - configure - Generates a Makefile for the current module
    - rebuild - Runs "clean", "configure" and "build" all at once
    - install - Install node development files for the specified node version.
    - list - Prints a listing of the currently installed node development files
    - remove - Removes the node development files for the specified version

node-gyp@9.4.0  /opt/homebrew/lib/node_modules/node-gyp
node@20.8.0
```

File: Makefile (hunk 1 above)
Comment (pokey) [reply Makefile]: done!

**Comment by pokey** [comment]
It appears to not be using the right node-gyp; see the line in your output where it says `gyp ERR! node-gyp -v v7.1.2`

**Comment by phillco** [comment]
Right, hence my question. :) It looks like it was pinned in the yarn lock file:

```
~/s/c/vscode-parse-tree (*130* 8.14 s)  (pokey/reinstate-local-web-tree-sitter-build|…) > rg node-gyp
yarn.lock
790:    node-gyp "^7.1.0"
1711:node-gyp@^7.1.0, node-gyp@^7.1.2:
1713:  resolved "https://registry.yarnpkg.com/node-gyp/-/node-gyp-7.1.2.tgz#21a810aebb187120251c3bcec979af1587b188ae"
2464:    node-gyp "^7.1.2"
~/s/c/vscode-parse-tree (115 ms)  (pokey/reinstate-local-web-tree-sitter-build|…) > rm yarn.lock
~/s/c/vscode-parse-tree (9 ms)  (pokey/reinstate-local-web-tree-sitter-build|✚1…) > yarn
<Lots of output that's still running>
```

**Comment by pokey** [comment]
Oh interesting. Strange that it works on mine, because I'm using the same yarn lock file. Could be that you're on a newer node version than I am; I'm on `v19.8.1`. Unfortunately, looks like tree-sitter-haskell [pins node-gyp](https://github.com/tree-sitter/tree-sitter-haskell/blob/d7ac98f49e3ed7e17541256fe3881a967d7ffdd3/package.json#L15) to 7.x.x 😕. So @phillco you might need to try building with an older node version. @auscompgeek @wenkokke any ideas on the way forward here? Do we file a PR against tree-sitter-haskell to bump their node-gyp dep?

**Comment by phillco** [comment]
Still fails under 19.8.1. Do you know what the right version of node-gyp is?

**Comment by phillco** [comment]
(This isn't super blocking for me btw so I wouldn't worry too much about it if this is a lot of work to fix.)

**Pull request merged**